crates/ere-risc0/src/prove/bento.rs

```diff
@@ -0,0 +1,230 @@
+        .map_err(|err| zkVMError::Other(err.into()))?;
+    let input_id = client
+        .upload_input(input_bytes)
+        .map_err(|err| zkVMError::Other(err.into()))?;
+
+    let now = std::time::Instant::now();
+
+    let session = client
+        .create_session(program.image_id.to_string(), input_id, vec![], false)
+        .map_err(|err| zkVMError::Other(err.into()))?;
+
+    loop {
+        let res = session
+            .status(&client)
+            .map_err(|err| zkVMError::Other(err.into()))?;
+
+        match res.status.as_ref() {
+            "RUNNING" => {
+                std::thread::sleep(Duration::from_millis(200));
+                continue;
+            }
+            "SUCCEEDED" => {
+                let receipt_bytes = client
+                    .receipt_download(&session)
+                    .map_err(|err| zkVMError::Other(err.into()))?;
+                break Ok((bincode::deserialize(&receipt_bytes).unwrap(), now.elapsed()));
+            }
+            _ => {
+                return Err(zkVMError::Other(
+                    format!("Unexpected proving status: {}", res.status).into(),
+                ));
+            }
+        }
+    }
+}
+
+fn cmd_output_checked(cmd: &mut Command) -> Result<Output, io::Error> {
+    let output = cmd.output()?;
+    if !output.status.success() {
+        return Err(io::Error::other(format!("Failed to run `{cmd:?}`")));
+    }
+    Ok(output)
+}
+
+fn cmd_exec_checked(cmd: &mut Command) -> Result<(), io::Error> {
+    let status = cmd.status()?;
+    if !status.success() {
+        return Err(io::Error::other(format!("Failed to run `{cmd:?}`")));
+    }
+    Ok(())
+}
+
+fn docker_image_exists(image: impl AsRef<OsStr>) -> Result<bool, io::Error> {
+    let output = cmd_output_checked(
+        Command::new("docker")
+            .args(["images", "--quiet"])
+            .arg(image),
+    )?;
+    // If image exists, image id will be printed hence stdout will be non-empty.
+    Ok(!output.stdout.is_empty())
+}
+
+fn docker_image_tag(src: impl AsRef<OsStr>, dst: impl AsRef<OsStr>) -> Result<(), io::Error> {
+    cmd_exec_checked(
+        Command::new("docker")
+            .args(["image", "tag"])
+            .arg(src)
+            .arg(dst),
+    )
+}
+
+pub fn build_bento_images() -> Result<(), io::Error> {
+    let agent_tag = format!("ere-risc0/agent:{SDK_VERSION}");
+    let rest_api_tag = format!("ere-risc0/rest_api:{SDK_VERSION}");
+
+    if docker_image_exists(&agent_tag)? && docker_image_exists(&rest_api_tag)? {
+        return Ok(());
+    }
+
+    let tempdir = tempdir()?;
+
+    cmd_exec_checked(
+        Command::new("git")
+            .args([
+                "clone",
+                "--single-branch",
+                "--branch",
+                &format!("v{SDK_VERSION}"),
+                "https://github.com/risc0/risc0.git",
+            ])
+            .arg(tempdir.path()),
+    )?;
+
+    cmd_exec_checked(
```

Comment:
Coffee talk: so if we're using `ere-risc0` natively and we prove multiple things then:
- Each time we'll pull the Risc0 git repo in a temp directory, and
- Build the Risc0 images

I think pulling the git repo with `--depth 0` should be okay. I just wonder if the `docker compose build` will leverage the docker cache. Mostly because the second, third, etc runs will always happen in a new folder (but the context should be the same).

Anyway, just a thought... not sure if you have seen this confirmed in runs. :)

Reply:
I think the only case we will build multiple time is by running `ere-risc0` (natively) or `ere-dockerized` risc0 (dockerized) for the first time, and instantiating multiple `EreRisc0` parallelly.

The reason is in line 113 it'd check if the image exist or not and early return, so ideally it'd only be built once. In my dev experience last few day I only need to build it once, and it took 20 mins so I guess I would have noticed if it builds again 🥲

But if in production env the local docker image gets lost every time, perhaps it'd be painful as you said.. In this case I guess it'd be worth to push this image to some registry to reuse it.